### PR TITLE
Add Logits tests for Qwen2.5VL

### PR DESCRIPTION
**Pull request by ravi03071991**
## Motivation

PR to add logits tests for Qwen2.5VL.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/utils.py [hunk 1]
```diff
@@ -808,19 +808,71 @@ def set_recv_opt():
     return socket
 
 
+# def dump_to_file(dirpath, name, value):
+#     from sglang.srt.distributed import get_tensor_model_parallel_rank
+
+#     if get_tensor_model_parallel_rank() != 0:
+#         return
```
Comment (yizhang2077) [inline python/sglang/srt/utils.py]: remove this comment

**Review by yizhang2077 — COMMENTED** [review]
It seems MiniCPMV test has been removed. Can we keep both of MiniCPMV and Qwen2.5VL tests?

File: python/sglang/srt/utils.py (hunk 1 above)
Comment (ravi03071991) [reply python/sglang/srt/utils.py:815]: oh yeah my bad. I will add them.

**Comment by ravi03071991** [comment]
> It seems MiniCPMV test has been removed. Can we keep both of MiniCPMV and Qwen2.5VL tests?

Yeah sure. Sorry, I removed it for local testing but added it back.

File: python/sglang/srt/utils.py (hunk 1 above)
Comment (ravi03071991) [reply python/sglang/srt/utils.py]: Done.

File: python/sglang/srt/utils.py [hunk 2]
```diff
@@ -809,18 +809,55 @@ def set_recv_opt():
...
     os.makedirs(dirpath, exist_ok=True)
+
     if value.dtype is torch.bfloat16:
         value = value.float()
     value = value.cpu().numpy()
-    output_filename = os.path.join(dirpath, f"pytorch_dump_{name}.npy")
-    logger.info(f"Dump a tensor to {output_filename}. Shape = {value.shape}")
-    np.save(output_filename, value)
+
+    output_filename = os.path.join(dirpath, f"pytorch_dump_{name}.npz")
```
Comment (yizhang2077) [inline python/sglang/srt/utils.py:930]: I think it is robust to compare logits in each step, but I don't know if this change may effect other people debug? @merrymercy . And I think seperate prefill/decode stages by file's existence is not a good way. Maybe passing `forward_mode` into this function is better.

Comment (ravi03071991) [reply python/sglang/srt/utils.py:930]: Logits are compared at each step. Since the first token is generated during the prefill stage, I have been saving the prefill logits as well. However, I could instead save only the logits of the generated token during the prefill stage.

Comment (ravi03071991) [reply python/sglang/srt/utils.py:930]: @yizhang2077 I think Grok model file uses [dump_to_file](https://github.com/sgl-project/sglang/blob/b36ab493b3ad6fbd1d663493cd2362a247c3308f/python/sglang/srt/models/grok.py#L232) for debugging purposes. I don’t think it’s the right approach to change this for obvious reasons. Wouldn’t it be better to explore an alternative solution? What do you think?

Comment (yizhang2077) [reply python/sglang/srt/utils.py:930]: @ravi03071991 Ok, I think it is fine, can you try to fix the conlicts? After that we can try to merge it~

**Comment by ravi03071991** [comment]
@yizhang2077 resolved the conflicts but yeah the logits tests are failing. Do you see something wrong here?

**Comment by yizhang2077** [comment]
Our test is not stable, I will rerun it @ravi03071991

**Review by yizhang2077 — APPROVED** [review]

**Comment by yizhang2077** [comment]
@ravi03071991 This test can not pass, would you like to check it?

**Comment by ravi03071991** [comment]
> @ravi03071991 This test can not pass, would you like to check it?

Seems like there is issue with HF generate or something. Let me check on it.

**Comment by yizhang2077** [comment]
@ravi03071991 This PR may affect test results, plz notice https://github.com/sgl-project/sglang/pull/4661

File: test/srt/test_vlm_accuracy.py [hunk 3]
```diff
@@ -211,5 +226,141 @@ async def test_vlm_embedding_output(self):
         self.compare_outputs(sglang_output, hf_output)
 
 
+class TestQWEN2VLLogits(VisionLLMLogitsBase):
```
Comment (Qiaolin-Yu) [inline test/srt/test_vlm_accuracy.py:229]: This might be a stupid question, but could this test use `SRTRunner` and `HFRunner`? Will it be simpler?

Comment (ravi03071991) [reply test/srt/test_vlm_accuracy.py:229]: I think we can, SRTRunner is a wrapper around `Engine` api and I dont see a direct way to get logits there? but I think current approach I feel simpler.

Comment (Qiaolin-Yu) [reply test/srt/test_vlm_accuracy.py:229]: `srt_outputs = srt_runner.forward(xxx)`

Then fetch `srt_outputs.top_input_logprobs`, `srt_outputs.top_output_logprobs`, etc.

Comment (ravi03071991) [reply test/srt/test_vlm_accuracy.py:229]: Okay. I am looking at comparing logits and not log probs. We can get log probs directly with Engine API.

Comment (ravi03071991) [reply test/srt/test_vlm_accuracy.py:229]: Or can we compare logprobs from srtrunner and hfrunner?

Comment (Qiaolin-Yu) [reply test/srt/test_vlm_accuracy.py:229]: OK. I may not have better ideas to fetch logits. But if you want `hidden_states` too, you can config             `return_hidden_states=True` when using generate API. Would logprobs + hidden_states be enough for you to compare?

Comment (ravi03071991) [reply test/srt/test_vlm_accuracy.py:229]: Let me try with logprobs once. Thanks @Qiaolin-Yu .

**Comment by merrymercy** [comment]
any new updates?

**Comment by ravi03071991** [comment]
> any new updates?

I had paused this while working on benchmark CI tasks. I’ll be resuming it and will share an update later this week.

**Comment by ravi03071991** [comment]
Closing this for now due to no progress on my end. Will revisit and reopen it at a later time.
